File: F#/chap06/Chap0605.fs
namespace Chap0605

module C060501 =
    type Order =
        { OrderLines: OrderLine list
          AmountToBill: float }

    and OrderLine = { id: int; Price: float }

    let findOrderLine orderLineId lines =
        lines |> List.find (fun ol -> ol.id = orderLineId)

    let replaceOrderLine orderLineId newOrderLine lines =
        lines |> List.map (fun ol -> if ol.id = orderLineId then newOrderLine else ol)


    let changeOrderLinePrice order orderLineId newPrice =
        let orderLine = order.OrderLines |> findOrderLine orderLineId
        let newOrderLine = { orderLine with Price = newPrice }
        let newOrderLines = order.OrderLines |> replaceOrderLine orderLineId newOrderLine
        let newAmountToBill = newOrderLines |> List.sumBy (fun line -> line.Price)

        { order with
            OrderLines = newOrderLines
            AmountToBill = newAmountToBill }

File: Scala/chap06/src/main/scala/Chap0605.scala
import scala.util.chaining.scalaUtilChainingOps

object Chap0605:
  case class OrderLine(id: Int, price: Float)

  case class Order(lines: Seq[OrderLine], amountToBill: Float)

  private def findOrderLine(orderLineId: Int, lines: Seq[OrderLine]): Either[String, OrderLine] =
    lines.find(_.id == orderLineId) match
      case Some(line) => Right(line)
      case None => Left(s"OrderLine with id $orderLineId not found")

  private def replaceOrderLine(orderLineId: Int, newLine: OrderLine, lines: Seq[OrderLine]): Seq[OrderLine] =
    lines.map { line => if line.id == orderLineId then newLine else line }

  def changeOrderLinePrice(order: Order, orderLineId: Int, newPrice: Float): Either[String, Order] =
    for {
      // Either lane
      orderLine <- findOrderLine(orderLineId, order.lines)
    } yield {
      orderLine.copy(price = newPrice)
        .pipe(replaceOrderLine(orderLineId, _, order.lines))
        .pipe(lines => Order(lines, lines.map(_.price).sum))
    }
